sdk/eventhub/Azure.Messaging.EventHubs/design/event-processor{T}-proposal.md

```diff
@@ -0,0 +1,520 @@
+
+- Allows for a custom set of information to be tracked and flow through the infrastructure for custom context associated with a specific partition.
+
+### Event Processor Options 
+
+In addition to the standard connection and retry options, allows configuration of lower-level details such as the batch size and prefetch count for optimal tuning.
+
+### Event Processor Partition
+
+Represents the base for custom partition context types used within the procesor to represent the active partition context for a given operation.
+
+### Event Processor Checkpoint
+
+The representation of a the last event seen by a processor for a given Event Hub, consumer group, and partition combination.  This is intended to be a durable representation of state, allowing processing to resume from the next event should partition ownership change or processing restart.
+
+### Partition Ownership
+
+Denotes that a specific Event Processor owns the responsibility of processing the partition for a given Event Hub and consumer group.  Ownership may change as processing work is balanced among active processors for the Event Hub and consumer group combination.
+
+## Usage examples
+
+### Create a minimal custom processor
+
+```csharp
+public class CustomProcessor : EventProcessor<EventProcessorPartition>
+{
+    // Local event processing
+    protected override async Task OnProcessingEventBatchAsync(IEnumerable<EventData> events, EventProcessorPartition partition, CancellationToken cancellationToken = default)
+    {
+        await DoSomethingWithEvents(events).ConfigureAwait(false);
+        Log.Information($"Received event batch for partition: { partition.PartitionId }");
+    }
+    
+    protected override Task OnProcessingErrorAsync(Exception exception, EventProcessorPartition partition, string operationDescription, CancellationToken cancellationToken = default)
+    {
+        Log.Error(exception, $"Error occured for partition: { partition.PartitionId } during { operationDescription }.");
+        return Task.CompletedTask;
+    }
+    
+    // Storage integration
+    protected override Task<IEnumerable<EventProcessorCheckpoint>> ListCheckpointsAsync(CancellationToken cancellationToken) =>
+        CustomStorage.GetCheckpointsAsync(FullyQualifiedNamespace, EventHubName, ConsumerGroup);
+    
+    protected override Task<IEnumerable<PartitionOwnership>> ListOwnershipAsync(CancellationToken cancellationToken) =>
+        CustomStorage.GetOwnershipAsync(FullyQualifiedNamespace, EventHubName, ConsumerGroup, Identifier);
+    
+    protected override Task<IEnumerable<PartitionOwnership>> ClaimOwnershipAsync(IEnumerable<PartitionOwnership> desiredOwnership, CancellationToken cancellationToken) =>
+        CustomStorage.TryUpdateOwnershipAsync(desiredOwnership);
```

Comment:
There is no more `UpdateCheckpoint` method, as the base doesn't call it.   Since developer code is responsible for calling it, they can use whatever signature and form they wish to create the checkpoints.   :smile: